rpm/src/main/java/org/eclipse/packager/rpm/parse/HeaderValue.java

```diff
@@ -89,84 +68,60 @@ public int getIndex() {
         return this.index;
     }
 
-    void fillFromStore(final ByteBuffer storeData) throws IOException {
+    @SuppressWarnings("unchecked")
```

Comment:
I don't think the idea of using `<T>` here works out. Due to the type erasure which will happen, this will be the same code as code (`Object`). And the cast of `RpmTagValue<T>` will never trigger any error, it's the same as `RpmTagValue<?>` IIRC.

Reply:
OK, you're right, nothing is really gained by using the `<E> Something<E>` pattern over `Something<?>`, but I am not sure what the difference is.

Reply:
I think using `Something<?>` clearly indicates that this is the case. Where `Something<E>` gives a false impression. Also, it might lead to class cast exceptions:

```java
public class Main {
    static <E> E convert(Object value) {
        return (E) value;
    }

    public static void main(String[] args) {
        boolean value = true;
        String converted = convert(value);
    }
}
```

The code will compile (with a warning the we suppress for `RpmTagValue`). However running the code will result in a `ClassCastException`. So it doesn't really add type safety, but creates a confusing API. 

To my my understanding, in our case this might lead to a call where the user expects to get a `RpmTagValue<String>`, but indeed it would be `RpmTagValue<Object>`, and the content might be `Boolean`. As some point, assigning that to a `String`, it would trigger a `ClassCastException`.

Reply:
Here's an example from the Android SDK, where they did the same, and fixed it later: https://developer.android.com/reference/android/content/Intent#getParcelableArrayListExtra(java.lang.String). The `ArrayList<T>` might have contained all kinds of objects, not just `T`. Simply because that information (of T) is not available during runtime.

They replaced it with a method which received `Class<T>`, and so it is possible (during runtime) to `.cast()` to `T`, as the class information is present.

Having a background in C++ and working with Rust, this all seems weird :) I personally think type erasure in Java was a big mistake.

Reply:
Yes, I was trying to use this pattern of passing the class for the `getOptionalTag`/`getEntry` API.